src/messaging/ExchangeMgr.h

```diff
@@ -0,0 +1,266 @@
+
+using System::PacketBuffer;
+
+class ExchangeContext;
+
+static constexpr int16_t ANY_MESSAGE_TYPE = -1;
+
+/**
+ *  @brief
+ *    This class is used to manage ExchangeContexts with other CHIP nodes.
+ *    It works on behalf of higher layers, creating ExchangeContexts and
+ *    handling the registration/unregistration of unsolicited message handlers.
+ */
+class DLL_EXPORT ExchangeManager : public SecureSessionMgrDelegate
+{
+public:
+    ExchangeManager();
+    ExchangeManager(const ExchangeManager &) = delete;
+    ExchangeManager operator=(const ExchangeManager &) = delete;
+
+    /**
+     *  Initialize the ExchangeManager object. Within the lifetime
+     *  of this instance, this method is invoked once after object
+     *  construction until a call to Shutdown is made to terminate the
+     *  instance.
+     *
+     *  @param[in]    sessionMgr    A pointer to the SecureSessionMgrBase object.
+     *
+     *  @retval #CHIP_ERROR_INCORRECT_STATE If the state is not equal to
+     *          kState_NotInitialized.
+     *  @retval #CHIP_NO_ERROR On success.
+     *
+     */
+    CHIP_ERROR Init(SecureSessionMgrBase * sessionMgr);
+
+    /**
+     *  Shutdown the ExchangeManager. This terminates this instance
+     *  of the object and releases all held resources.
+     *
+     *  @note
+     *     The application should only call this function after ensuring that
+     *     there are no active ExchangeContext objects. Furthermore, it is the
+     *     onus of the application to de-allocate the ExchangeManager
+     *     object after calling ExchangeManager::Shutdown().
+     *
+     *  @return #CHIP_NO_ERROR unconditionally.
+     */
+    CHIP_ERROR Shutdown();
+
+    /**
+     *  Creates a new ExchangeContext with a given peer CHIP node specified by the peer node identifier.
+     *
+     *  @param[in]    peerNodeId    The node identifier of the peer with which the ExchangeContext is being set up.
+     *
+     *  @param[in]    appState      A pointer to a higher layer object that holds context state.
+     *
+     *  @return   A pointer to the created ExchangeContext object On success. Otherwise NULL if no object
+     *            can be allocated or is available.
+     */
+    ExchangeContext * NewContext(const uint64_t & peerNodeId, void * appState = nullptr);
+
+    /**
+     *  Find the ExchangeContext from a pool matching a given set of parameters.
+     *
+     *  @param[in]    peerNodeId    The node identifier of the peer with which the ExchangeContext has been set up.
+     *
+     *  @param[in]    appState      A pointer to a higher layer object that holds context state.
+     *
+     *  @param[in]    isInitiator   Boolean indicator of whether the local node is the initiator of the exchange.
+     *
+     *  @return   A pointer to the ExchangeContext object matching the provided parameters On success, NULL on no match.
+     */
+    ExchangeContext * FindContext(uint64_t peerNodeId, void * appState, bool isInitiator);
```

Comment:
Is this a public method? Which layer would call it?

Reply:
Yes, it is used by the protocol layer to search for an existing exchange for a peer node.